### PR TITLE
2087 allow when clause without instruction and forbid multiple when

### LINKED ISSUE #2087: Search all : allow when clause without instruction and forbid multiple when
**Describe the bug**
Search all : 
- allow when clause without instruction
- forbid more than 1 when clause

**To Reproduce**
Cobol code that cause the bug : (if any)
```cobol
       IDENTIFICATION DIVISION.
       PROGRAM-ID. SearchAll.
       data division.
       working-storage section.
       01 tab.
          05 var1 occurs 10 ascending key is Var2 indexed by idx .
            10 var2 pic X(1) .
      
       PROCEDURE DIVISION.    
      
      *-------- ONLY WHEN
      *OK, when with instruction
           search all var1
              when var2(idx) = 'b'
                display "test"
           end-search
      
      *OK, when without instruction
           search all var1
              when var2(idx) = 'b'
      
           end-search
      
      *-------- AT END and WHEN
      *OK
           search all var1
              at end
                display "end"
              when var2(idx) = 'b'
                display "when"
           end-search
      
      *OK
           search all var1
              at end
                display "end"
              when var2(idx) = 'b'
           end-search
      
      
      *-------- AT END at wrong position
           search all var1
              when var2(idx) = 'b'
                display "when"
      *KO at end must be before when statement
              at end
                display "end"
           end-search
      
           search all var1
              when var2(idx) = 'b'
      *KO at end must be before when statement
              at end
                display "end"
           end-search

      *-------- 2 WHEN always KO
           search all var1
              when var2(idx) = 'a'
                display "when a"
              when var2(idx) = 'b'
                display "when b"
           end-search
      
           search all var1
              at end
                display "end"
              when var2(idx) = 'a'
                display "when a"
              when var2(idx) = 'b'
                display "when b"
           end-search
      
           search all var1
              at end
                display "end"
              when var2(idx) = 'a'
              when var2(idx) = 'b'
                display "when a and b"
           end-search
           goback.
       END PROGRAM SearchAll.
```

**Expected behavior**
Report only expected error in above source code.

### PR DESCRIPTION
**Pull request by laabidiHend**
Fixes #2087 #2088

### CODE REVIEW

File: TypeCobol/Compiler/CupParser/TypeCobolProgram.cup [hunk 1]
```diff
@@ -1449,7 +1449,11 @@ searchStatement ::= SearchStatement:stmt
 	;
 
 whenSearchConditionClause ::= 
-		whenSearchCondition  statements
+        whenSearchCondition
```
Comment (fm-117) [inline TypeCobol/Compiler/CupParser/TypeCobolProgram.cup:1452]: indent

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs [hunk 2]
```diff
@@ -310,11 +350,18 @@ public override bool Visit(Search search)
...
+        
         public override bool Visit(WhenSearch whenSearch)
         {
             System.Diagnostics.Debug.Assert(whenSearch.Parent is Search);
             var search = (Search) whenSearch.Parent;
 
+            if (whenSearch.ChildrenCount == 0)
+            {
+                var messageCode = search.CodeElement.StatementType == StatementType.SearchSerialStatement ? MessageCode.SyntaxErrorInParser : MessageCode.Warning;
+                DiagnosticUtils.AddError(whenSearch, "Missing statement in when clause",messageCode);
```
Comment (fm-117) [inline TypeCobol/Compiler/Diagnostics/CrossChecker.cs:362]: Missing space after second ','

**Review by fm-117 — CHANGES_REQUESTED** [review]
This PR fails at build because of the 2 tests taht we attempted to correct. @laabidiHend, please restore the diagnostics in WhenTest.rdzMix.txt and WhenTestNoendEvaluate.rdzMix.txt. Also don't forget to update your branch with develop.

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs (hunk 2 above)
Comment (laabidiHend) [reply TypeCobol/Compiler/Diagnostics/CrossChecker.cs]: Fixes  ffc24a263c66ce8beb0feeb493fb5572060ccf59 2e092eb2943043ef771dbc24b9789b818a240caf

**Review by smedilol — CHANGES_REQUESTED** [review]

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs [hunk 3]
```diff
@@ -254,10 +254,50 @@ public override bool Visit(Move move)
...
+            bool onAtEndExist = false;
+            int index = 0;
+            foreach (var child in search.Children)
+            {
+                if (child is OnAtEnd)
+                {
+                    if (index > 0)
+                    {
+                        DiagnosticUtils.AddError(child, "ON AT END clause must appear before WHEN.");
+                    }
```
[the remark this answers is not in the document]
Comment (laabidiHend) [reply TypeCobol/Compiler/Diagnostics/CrossChecker.cs]: i did the changes requested 2ca312a4dba3b9cd39cdcb14a394992afbbd1f32

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs [hunk 4]
```diff
@@ -254,10 +254,50 @@ public override bool Visit(Move move)
...
+                        DiagnosticUtils.AddError(child, "ON AT END clause must be unique.");
+                    }
+                    else
+                    {
+                        onAtEndExist = true;
+                    }
+                }
+                else if (child is WhenSearch whenSearch)
+                {
+                    list.Add(whenSearch);
```
[the remark this answers is not in the document]
Comment (laabidiHend) [reply TypeCobol/Compiler/Diagnostics/CrossChecker.cs]: Done 2ca312a4dba3b9cd39cdcb14a394992afbbd1f32

**Review by fm-117 — APPROVED** [review]

**Review by smedilol — DISMISSED** [review]

**Review by mayanje — APPROVED** [review]

**Review by fm-117 — APPROVED** [review]

**Pull request merged**